### ISSUE TITLE
App becomes unresponsive and presents ANR when tapping "On Device"

### ISSUE THREAD

**Issue by mrhebisan** [nextcloud/android#8074]
### Steps to reproduce

On a Pixel 4a, 4G version, 3a, and Samsung s10e

1. Create Folders Music, and Movies on a NextCloud install on a VPS
2. Sync 8GB of music and 1GB of video to the VPS
3. Install the apk from F-Droid on your phone and leave all settings as default
4. Log into Nextcloud using the app
5. Sync your Music and Movies folder to device
6. Click the "on device" menu item in the Next cloud app

### Expected behaviour
- I see the On Device screen

### Actual behaviour
- The app becomes completely unresponsive as the UI thread hangs.  After several seconds the app shows the ANR dialog allowing you to Force Quit or Wait

### Can you reproduce this problem on https://try.nextcloud.com?
I would rather not post my music and videos to your account

### Environment data
Android version: 11

Device model: 
Pixel 4a (4G version)
Pixel 3a
Samsung s10e (model SM-G970U1)

Stock or customized system: various

Samsung s10e : Unmodified.  
Google play system update Feb 1 2021
Build RP1A.200720.012.G970U1UEU4FUA4

Pixel 4a: GrapheneOS ROM, latest stable version Build RQ1A.210205.004.2021.02.19.15

Pixel 3a: GrapheneOS ROM, latest stable version
Build RQ1A.210205.004.2021.02.19.15

Nextcloud app version: 3.15.0

Nextcloud server version: 21.0.0 zip from the nextcloud site

Reverse proxy: ???

#### Web server error log
```
Where can I find them??
```

#### Nextcloud log (data/nextcloud.log)
```
Contains data I prefer to remain private
```
**NOTE:** Be super sure to remove sensitive data like passwords, note that everybody can look here! You can use the Issue Template application to prefill some of the required information: https://apps.nextcloud.com/apps/issuetemplate

**Comment by tobiasKaminsky** [comment]
Is there still some sync/downl3oading done?
Or does this also happen if you start the app and then go directly to "on device"?

**Comment by mrhebisan** [comment]
The issue persists even without any active sync in process (that is apparent at least).  The syncs completed days ago but I can reproduce the issue still on all the devices listed.  The devices take over 30 seconds to display their respective screens, and an ANR popup can result.

The issue also emerges if the app is minimized while in the On Device screen and then brought back to the foreground, as the app takes over 30 seconds to display the list.

**Comment by tobiasKaminsky** [comment]
how many downloaded files / folders do you have?

**Comment by tobiasKaminsky** [comment]
If I remember correctly we compute local stored size for each folder, and maybe this takes too much time.

**Comment by mrhebisan** [comment]
Under music
 73 bands with a folder per band
 At least 1-4 album folders per band
 1-20 songs per album
 Maybe 100 songs in various artists
 An iTunes folder with all the bands from my iTunes downloads 

Under Movies
  About 5 movies

The problem doest get worse the more files and folders in music. I got alphabetically to the iTunes folder and had counted 390 songs.  There are probably almost 2000 files at least.

**Comment by tobiasKaminsky** [comment]
We indeed compute entire folder size for each folder when listing local files:

https://github.com/nextcloud/android/blob/a42e0b42f4adef24a72371ebd9c85df9b837a48b/src/main/java/com/owncloud/android/utils/FileStorageUtils.java#L286

**Comment by tobiasKaminsky** [comment]
Maybe we should offload this computation to a thread, so that it does not delay showing files, and show meanwhile "computing…".
@AndyScherzinger

**Comment by mrhebisan** [comment]
Could the local folder sizes be cached, and only recomputed if the local file system has changed or updated?

**Comment by AndyScherzinger** [comment]
> Maybe we should offload this computation to a thread, so that it does not delay showing files, and show meanwhile "computing…". @AndyScherzinger

Yes, that sound like a practical and good idea

> Could the local folder sizes be cached, and only recomputed if the local file system has changed or updated?

This might fix is in cases where it didn't change but leads to the same problem on a change, so doing this asynchronously is the way to generally solve it no matter the scenario.

**Comment by tobiasKaminsky** [comment]
There is an APK file in PR #8130, which you can install in parallel to your existing Nextcloud app.
It would be great if you can verify the fix.

@mrhebisan if you prefer to test it with your actual app, I can also create you an upgrade.

**Comment by mrhebisan** [comment]
Thank you, I'll try side by side for now. I've got the free space to spare currently.

May take a little time to try on all 3 devices, but already I noticed one issue:

If I play a music file that has finished syncing while other files sync, it appears to stop all in-progress sync. Same thing when I background the app in order to write this reply.

Seems the sync process doesn't hold up when backgrounding on stock Samsung. I see a series of failed to download notifications and then finally on attempt that is stuck at 0%

Sent from ProtonMail mobile

-------- Original Message --------

**Comment by mrhebisan** [comment]
It is possible to navigate around "On Device" a bit now, and I do see calculating messages for folder sizes.  

The problem can still be reproduced while exploring "On Device" but with less clear reproduction steps on all devices listed.  Could it be related to how deeply nested the songs are inside the folder structures?  I do have to quickly navigate up and down many levels to explore the file system. 

 Further, when the application is too busy various files may fail to download.  When it finally catches up, I have almost 50 notifications for failed downloads.

Another bug this introduces is that if I play a song that I've successfully synced while this process is going wrong, there is no way for me to stop the audio once I leave the audio player NextCloud opens.  In fact, I closed all running applications and the audio still did not stop, and I was unable to bring up any player that had control of it.  On the whole, the update introduces more issues than it resolves

**Comment by tobiasKaminsky** [comment]
Please only one bug in one issue :-)

Regarding this and APK in #8130, it is only about computing.
It might be that there are too many computing backgrounds, when you navigate too quickly.
Then I'll have to stop them, once they are not valid anymore.

I currently fail to understand what the download notification has to do with it?
I thought it is "just" about app getting unresponsible when you browse in "on device"?

**Comment by mrhebisan** [comment]
The other issues aside since they only occur in #8130. I'll put them aside for now.

"It might be that there are too many computing backgrounds, when you navigate too quickly. Then I'll have to stop them, once they are not valid anymore."

That seems reasonable.

"I currently fail to understand what the download notification has to do with it?

I thought it is "just" about app getting unresponsible when you browse in "on device"?"

It seems that navigating the folders quickly while a large sync is in process can cause this issue in #8130. It indicates that navigating the folders causes some file synch failure. If that is out of scope for #8130 or If that is expected behavior then let me know.

Sent with [ProtonMail](https://protonmail.com) Secure Email.

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐

**Comment by tobiasKaminsky** [comment]
> It seems that navigating the folders quickly while a large sync is in process can cause this issue in #8130.

You browse this folders in "on device" view, or? Then there should be no sync, but only shows all existing files/folders, if I remember correctly.
